File: src/ReactWidget.java
package org.proactive;

import org.proactive.prolog.PrologObject;
import org.proactive.prolog.PrologState;
import org.proactive.prolog.Engine;
import org.proactive.prolog.FluxDispatcher;
import org.proactive.ui.ProactiveConstraints;
import gnu.prolog.term.Term;
import gnu.prolog.term.CompoundTerm;
import gnu.prolog.term.AtomTerm;
import gnu.prolog.vm.TermConstants;
import gnu.prolog.vm.PrologException;
import java.util.List;
import java.util.Stack;
import java.util.LinkedList;
import java.util.HashMap;
import java.awt.Component;
import java.awt.GridBagConstraints;
import javax.swing.JPanel;
import javax.swing.SwingUtilities;
import java.awt.BorderLayout;

/* Important notes
 *    o A ReactWidget does NOT contain any children!
 *    o it DOES however, have a visual representation. Imagine it as a photo of another component
 *    o It has a single reference internalComponent which has its own life. We just grab the AWT component from it when rendering
 */


public class ReactWidget extends ReactComponent
{
   protected Engine engine;
   protected String elementId;
   protected PrologState state;
   protected PrologState props;
   protected Term vDom = null;
   private boolean hasFluxListeners = false;
   private static int globalWidgetId = 0;
   private String widgetId;
   private ReactComponent internalComponent;
   private Term listeners = TermConstants.emptyListAtom;
   public ReactWidget(ReactWidget parentContext, Engine engine, String elementId, PrologState props) throws Exception
   {
      this.engine = engine;
      this.elementId = elementId;
      this.props = props;
      this.owner = parentContext;
      this.widgetId = "$widget" + (globalWidgetId++);
      engine.registerWidget(this);
      setProperties(props.getProperties());

      // First, get the state
      this.state = engine.getInitialState(elementId, props);
      // Then render the initial vDOM
      vDom = engine.render(this, elementId, state, props);

      // Now turn that into an actual component
      internalComponent = engine.createElementFromVDom(vDom, this);
      internalComponent.setOwnerDocument(this);

      // Next, check to see if the module needs any flux listeners
      hasFluxListeners = engine.checkForFluxListeners(this);
      listeners = engine.checkForMessageHandlers(this, listeners);
   }

   public String getWidgetId()
   {
      return widgetId;
   }

   public Component getAWTComponent()
   {
      return internalComponent.getAWTComponent();
   }

   public void setParentNode(ReactComponent parent)
   {
      super.setParentNode(parent);
      // Destroy child if we are being destroyed
      if (parent == null && internalComponent != null)
      {
         destroy();
         internalComponent.setParentNode(null);
      }
   }

   public String toString()
   {
      return "<Widget:" + elementId + " " + props+ ">";
   }

   public ProactiveConstraints.Fill getFill()
   {
      if (internalComponent != null)
         return internalComponent.getFill();
      return super.getFill();
   }

   @Override
   public ProactiveConstraints.Alignment getSelfAlignment()
   {
      if (internalComponent != null)
         return internalComponent.getSelfAlignment();
      return super.getSelfAlignment();
   }

   public PrologState getState()
   {
      return state;
   }

   public PrologState getProps()
   {
      return props;
   }

   public String getComponentName()
   {
      return elementId;
   }

   public void setState(PrologState newState) throws PrologException
   {
      state = newState;
      //System.out.println("State of " + elementId + " is now " + newState);
      listeners = engine.checkForMessageHandlers(this, listeners);
      reRender();
   }

   public void reRender() throws PrologException
   {
      Term newvDom = engine.render(this, elementId, state, props);
      Term patches = engine.diff(vDom, newvDom);
      //System.out.println("new VDOM: " + newvDom);
      internalComponent = engine.applyPatch(patches, internalComponent);
      internalComponent.setOwnerDocument(this);
      getAWTComponent().validate();
      vDom = newvDom;
   }

   public void destroy()
   {
      if (hasFluxListeners)
	 engine.deregisterFluxListener(elementId, this);
   }

   public ReactWidget updateWidget(PrologState newProps) throws PrologException
   {
      props = newProps;
      if (!engine.componentWillReceiveProps(elementId, this))
         reRender();
      return this;
   }


   public boolean triggerEvent(Term handler, Term event) throws PrologException
   {
      return engine.triggerEvent(handler, event, this);
   }

   public boolean triggerTest(Term handler, Term event) throws PrologException
   {
      return engine.triggerTest(handler, event, this);
   }


   public ReactComponent renderContextualElement(Term handler, Term event) throws PrologException
   {
      Term elementDom = engine.renderContextualElement(handler, event, this);
      ReactComponent element = engine.createElementFromVDom(elementDom, this);
      element.setOwnerDocument(this);
      return element;
   }


   public boolean fluxEvent(Term goal, String storeName, PrologState storeState) throws Exception
   {
      return engine.fluxEvent(goal, storeName, storeState, this);
   }

   public Engine getEngine()
   {
      return engine;
   }

   public ReactWidget getParentContext()
   {
      return owner;
   }
}

File: src/react_widget.js
"use strict";

var ReactComponent = require('./react_component');
var Prolog = require('proscript');
var Constants = require('./constants');
var global_widget_id = 0;

function ReactWidget(parentContext, engine, elementId, props, callback)
{
    ReactComponent.call(this);
    this.engine = engine;
    // Suppose a widget fires an event which results in the widget no longer existing.
    // Once the handler has completed, it will try and set the state on the (long since cleaned-up) widget
    // which will result in freeing (a second time) the state, and therefore a mess
    // To manage this, we have an explicit variable here: this.destroyed, which is set to true to indicate that
    // setState should be a NOP since the widget has been destroyed
    this.destroyed = false;
    this.elementId = elementId;
    this.props = props;
    this.owner = parentContext;
    this.vDom = null;
    this.internalComponent = null;
    this.id = "$widget" + (global_widget_id++);
    engine.registerWidget(this);
    this.setProperties(props.getProperties());
    this.listeners = Constants.emptyListAtom;
    engine.getInitialState(this, elementId, props, function(state)
                           {
                               this.state = state;
                               this.engine.render(this, this.elementId, this.state, this.props, function(vDom)
                                                  {
                                                      this.vDom = vDom;
                                                      engine.createElementFromVDom(this.vDom, this, function(internalComponent)
                                                                                   {
                                                                                       this.internalComponent = internalComponent;
                                                                                       this.internalComponent.setOwnerDocument(this);
                                                                                       //this.hasFluxListeners = engine.checkForFluxListeners(this);
                                                                                       this.internalComponent.restyle();
                                                                                       this.engine.checkForMessageHandlers(this,
                                                                                                                           function()
                                                                                                                           {
                                                                                                                               callback(this);
                                                                                                                           }.bind(this))
                                                                                       //callback.bind(this)(this);
                                                                                   }.bind(this));
                                                  }.bind(this));
                           }.bind(this));
}

ReactWidget.prototype = new ReactComponent;

ReactWidget.prototype.getDOMNode = function()
{
    if (this.internalComponent == null)
        return null;
    return this.internalComponent.getDOMNode();
}

ReactWidget.prototype.getEngine = function()
{
    return this.engine;
}

ReactWidget.prototype.getState = function()
{
    return this.state;
}

ReactWidget.prototype.getProps = function()
{
    return this.props;
}

ReactWidget.prototype.getComponentName = function()
{
    return this.elementId;
}

ReactWidget.prototype.setState = function(newState, callback)
{
    if (this.destroyed)
    {
        callback();
        return;
    }
    if (this.state != undefined && !this.state.is_global)
    {
        Prolog._release_blob("state", this.state.blob);
    }
    this.state = newState;
    this.engine.checkForMessageHandlers(this,
                                        function()
                                        {
                                            this.reRender(callback);
                                        }.bind(this))
}

ReactWidget.prototype.updateWidget = function(newProps, callback)
{
    if (this.props != undefined)
        this.props.freeState();
    this.props = newProps;
    this.engine.componentWillReceiveProps(this.elementId, this,
                                          function(didReceive)
                                          {
                                              if (!didReceive)
                                                  this.reRender(callback);
                                              else
                                                  callback(this);
                                          }.bind(this));
}

ReactWidget.prototype.destroyWidget = function(vNode)
{
    console.log("Destroy component: " + Prolog._portray(vNode));
    this.engine.deregisterWidget(this);
    this.freeComponent(vNode);
}

ReactWidget.prototype.freeComponent = function(vNode)
{
    this.engine.deregisterWidget(this);
    this.state.freeState();
    this.props.freeState();
    Prolog._release_blob("react_component", this.blob);
    this.destroyed = true;
    ReactComponent.prototype.freeComponent.call(this, vNode); //ie super.freeComponent(vNode)
}

ReactWidget.prototype.reRender = function(callback)
{
    this.engine.render(this, this.elementId, this.state, this.props, function(newVDom)
                       {
                           this.engine.diff(this.vDom, newVDom, function(patches)
                                            {
                                                this.engine.applyPatch(patches, this.internalComponent, function()
                                                                       {
                                                                           this.internalComponent.setOwnerDocument(this);
                                                                           if (this.vDom != null)
                                                                               Prolog._free_local(this.vDom);
                                                                           this.vDom = newVDom;
                                                                           //console.log("All patches applied");
                                                                           callback(this);
                                                                       }.bind(this));
                                            }.bind(this));
                       }.bind(this));
}

ReactWidget.prototype.triggerEvent = function(handler, event, callback)
{
    this.engine.triggerEvent(handler, event, this, callback);
}

ReactWidget.prototype.queueEvent = function(handler, event, callback)
{
    this.engine.queueEvent(handler, event, this, callback);
}


ReactWidget.prototype.debugStuff = function()
{
    this.engine.debugStuff();
}

module.exports = ReactWidget;
